### PR TITLE
[@begin/data] Add types.

### PR DESCRIPTION
**Pull request by LumaKernel**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
  - https://github.com/beginner-corp/begin-data/pull/85#issuecomment-881996721
  - Official package want not to include types in their repository.
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition: N/A

If removing a declaration: N/A

### CODE REVIEW

File: types/begin__data/begin__data-tests.ts [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    res[0].table = 'another-table';
+    // $ExpectType string
+    res[0].key;
+    // $ExpectType TestData
+    res[0] as TestData;
+    const cursor = res.cursor;
+    if (typeof cursor === 'string') {
+        // $ExpectType string
+        cursor;
+    } else {
```
Comment (peterblazejewicz) [inline types/begin__data/begin__data-tests.ts:45]:
```suggestion
    if (typeof cursor !== 'string') {
```

**Comment by peterblazejewicz** [comment]
You could be more explicit with `// $ExpectError`, this could remove the TS 3.9 dependency (if that was the only reason of the version freeze)

File: types/begin__data/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": ["es2018"],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictNullChecks": true,
+        "strictFunctionTypes": true,
+        "esModuleInterop": true,
```
Comment (peterblazejewicz) [inline types/begin__data/tsconfig.json:9]:
```suggestion
```
it's not (usually) used (or recommended) on DT and the package looks to be plain, old CJS:
https://github.com/beginner-corp/begin-data/blob/master/package.json#L5
https://github.com/beginner-corp/begin-data/blob/master/index.js#L9

File: types/begin__data/begin__data-tests.ts (hunk 1 above)
Comment (LumaKernel) [reply types/begin__data/begin__data-tests.ts:45]: I rewrote them with single line `// $ExpectType string | undefined` that is the same in the section below that.

File: types/begin__data/tsconfig.json (hunk 2 above)
Comment (LumaKernel) [reply types/begin__data/tsconfig.json:9]: Thank you! I've changed to use `import {set, get, ...} from ...` instead with this change.

**Comment by LumaKernel** [comment]
> You could be more explicit with // $ExpectError, this could remove the TS 3.9 dependency (if that was the only reason of the version freeze)

Thank you! I got the point of benefits of using `// $ExpectError` and I replace the minimum version comment with `3.7` that is actual minimum version that has first time supported recursively defined types.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@LumaKernel thx!

**Comment by LumaKernel** [comment]
Ready to merge

**Pull request merged**